src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs

```diff
@@ -2400,11 +2400,21 @@ public RuntimeTypeValue(TypeDesc type)
 
             public override bool GetRawData(NodeFactory factory, out object data)
             {
-                data = null;
-                return false;
+                data = factory.SerializedMaximallyConstructableRuntimeTypeObject(TypeRepresented);
```

Comment:
We are calling SerializedMaximallyConstructableRuntimeTypeObject here without taking `_devirtualizationManager.CanConstructType` info into account. The other places where this is called takes `_devirtualizationManager.CanConstructType` into account.

Can this disparity cause problems?

Reply:
It won't cause problems because we consistently ask for the same thing between scanning and codegen in the preinit interpreter.

We only go through the CanConstructType path when we need to downgrade `Type` used in comparisons. Whether downgrade is allowed is detected in the IL scanner when it sees the `==` pattern - scanner will report such Types as unconstructed. Then codegen will get the right form when it asks for a frozen version.

We'd need a lot more infrastructure to be able to do this downgrade for `Type` that was roundtripped through a static field (which is the path that's relevant here). When we're running the cctor interpreter, we don't know how the preinitialized static fields are going to be used: we'll only know if it's okay to downgrade once we know all the places that read from the field (they have to be in the form of the `==` pattern), including reflection access to the field. It's doable, but doesn't feel worth the trouble. Asking for the constructed version is the safe but not optimal option.

Reply:
Ok, thanks for the explanation.